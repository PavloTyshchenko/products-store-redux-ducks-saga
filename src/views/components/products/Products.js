import React from 'react';
import { connect } from 'react-redux';

import { CardColumns, Alert, Container } from 'react-bootstrap';

import ProductItem from './ProductItem';
import Spinner from '../spinner/Spinner';

const Products = (props) => {

    const {
        products,
        currentProducts,
        loading,
        showClear
    } = props;

    if (loading) {
        return (
            <Container className="text-center">
                <Spinner />
            </Container>
        );
    };

    let items = null;
    showClear ? items = currentProducts : items = products; 

    if (showClear && currentProducts.length === 0 && !loading) {
        return (
            <Alert variant='secondary'>
                Unfortunately no matches ...
            </Alert>
        );
    };

    return (
        <CardColumns className="mb-3">
            {
                items.map((product) => {
                    return <ProductItem key={product.name} product={product} />
                })
            }
        </CardColumns>
    );
};

const mapStateToProps = (state) => {
    return {
        products: state.product.products,
        currentProducts: state.product.currentProducts,
        loading: state.product.loading,
        showClear: state.product.showClear
    };
};

export default connect(mapStateToProps)(Products);
