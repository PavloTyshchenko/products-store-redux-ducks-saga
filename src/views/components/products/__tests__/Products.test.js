import React from 'react';
import { shallow } from 'enzyme';

import {Products} from '../Products';

const defaultProps = {
    products: [],
    category: 'All',
    search_term: '',
    loading: false,
    showClear: false
};


const setup = (propsOption) => {
    const props = Object.assign(defaultProps, propsOption);
    const wrapper = shallow(<Products {...props} />);
    return {
        wrapper,
        props
    };
};

describe('Products Component', () => {

    it('Should render without errors', () => {
        const { wrapper } = setup();
        const component = wrapper.find(`[data-test='productsComponent']`);
        expect(component.length).toBe(1);
    });

    it('Should render loading when loading is true', () => {
        const { wrapper } = setup({ loading: true });
        const component = wrapper.find(`[data-test='spinner']`);
        expect(component.length).toBe(1);
    });

    it('Should render alert when showClear is true, products length is 0 and loading is false', () => {
        const { wrapper } = setup({ showClear: true, loading: false  });
        const component = wrapper.find(`[data-test='alert']`);
        expect(component.length).toBe(1);
    });

    test('Should match snapshot', () => {
        const { wrapper } = setup({ products: [ { name: '1'}, { name: '2'}], showClear: false, loading: false});
        expect(wrapper.debug()).toMatchSnapshot();
    });
});