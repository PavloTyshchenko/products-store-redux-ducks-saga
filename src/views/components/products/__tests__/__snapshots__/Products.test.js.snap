// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Component Should match snapshot 1`] = `
"<CardColumns className=\\"mb-3\\" data-test=\\"productsComponent\\">
  <ProductItem product={{...}} />
  <ProductItem product={{...}} />
</CardColumns>"
`;
